lib/linkerd2-trace-context/src/layer.rs

```diff
@@ -0,0 +1,183 @@
+        }
+    }
+}
+
+// === impl Stack ===
+
+impl<T, M, S> tower::Service<T> for Stack<M, S>
+where
+    M: tower::Service<T>,
+    S: Clone,
+{
+    type Response = Service<M::Response, S>;
+    type Error = M::Error;
+    type Future = MakeFuture<M::Future, S>;
+
+    fn poll_ready(&mut self) -> Poll<(), M::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, target: T) -> Self::Future {
+        let inner = self.inner.call(target);
+
+        MakeFuture {
+            inner,
+            sink: Some(self.sink.clone()),
+        }
+    }
+}
+
+// === impl MakeFuture ===
+
+impl<F: Future, S> Future for MakeFuture<F, S> {
+    type Item = Service<F::Item, S>;
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let inner = try_ready!(self.inner.poll());
+        let sink = self.sink.take().expect("poll called after ready");
+        Ok(Async::Ready(Service { inner, sink }))
+    }
+}
+
+// === impl Service ===
+
+impl<Svc, B, S> tower::Service<http::Request<B>> for Service<Svc, S>
+where
+    Svc: tower::Service<http::Request<B>>,
+    S: SpanSink + Clone,
+{
+    type Response = Svc::Response;
+    type Error = Svc::Error;
+    type Future = ResponseFuture<Svc::Future, S>;
+
+    fn poll_ready(&mut self) -> Poll<(), Svc::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, mut request: http::Request<B>) -> Self::Future {
+        let trace_context = propagation::unpack_trace_context(&request);
+        let mut span = None;
+
+        if let Some(context) = trace_context {
+            trace!(message = "got trace context", ?context);
+            let span_id = propagation::increment_span_id(&mut request, &context);
+            // If we plan to sample this span, we need to record span metadata
+            // from the request before dispatching it to inner.
+            if context.is_sampled() {
+                trace!(message = "span will be sampled", ?span_id);
+                let path = request
+                    .uri()
+                    .path_and_query()
+                    .map(|pq| pq.as_str().to_owned());
+                span = Some(Span {
+                    trace_id: context.trace_id,
+                    span_id,
+                    parent_id: context.parent_id,
+                    span_name: path.unwrap_or_default(),
+                    start: SystemTime::now(),
+                    // End time will be updated when the span completes.
+                    end: SystemTime::UNIX_EPOCH,
+                });
+            }
+        }
+
+        let f = self.inner.call(request);
+
+        if let Some(span) = span {
+            ResponseFuture {
+                span: Some(span),
+                inner: f,
+                sink: Some(self.sink.clone()),
+            }
+        } else {
+            ResponseFuture {
+                span: None,
+                inner: f,
+                sink: None,
+            }
```

Comment:
i think i agree with eliza's suggestion to store sink/span together... but if not that, this can be simplified as:

```rust
let sink = span.as_ref().map(|_| self.sink.clone());
let inner = self.inner.call(req);
Response Future { inner, span, sink }
```